server/src/utils.ts

```diff
@@ -74,21 +76,72 @@ export let findProjectRootOfFile = (
   }
 };
 
-// Check if binaryName exists inside binaryDirPath and return the joined path.
-export let findBinary = (
-  binaryDirPath: p.DocumentUri | null,
-  binaryName: string
-): p.DocumentUri | null => {
-  if (binaryDirPath == null) {
+// If ReScript < 12.0.0-alpha.13, then we want `{project_root}/node_modules/rescript/{c.platformDir}/{binary}`.
+// Otherwise, we want to dynamically import `{project_root}/node_modules/rescript` and from `binPaths` get the relevant binary.
+// We won't know which version is in the project root until we read and parse `{project_root}/node_modules/rescript/package.json`
+let findBinary = async (
+  projectRootPath: p.DocumentUri | null,
+  binary: "bsc.exe" | "rescript-editor-analysis.exe" | "rescript"
+) => {
+  if (config.extensionConfiguration.platformPath != null) {
+    return path.join(config.extensionConfiguration.platformPath, binary);
+  }
+
+  const rescriptDir = lookup.findFilePathFromProjectRoot(
+    projectRootPath,
+    path.join("node_modules", "rescript")
+  );
+  if (rescriptDir == null) {
     return null;
   }
-  let binaryPath: p.DocumentUri = path.join(binaryDirPath, binaryName);
-  if (fs.existsSync(binaryPath)) {
-    return binaryPath;
+
+  let rescriptVersion = null;
+  let rescriptJSWrapperPath = null
+  try {
+    const rescriptPackageJSONPath = path.join(rescriptDir, "package.json");
+    const rescriptPackageJSON = JSON.parse(await fsAsync.readFile(rescriptPackageJSONPath, "utf-8"));
+    rescriptVersion = rescriptPackageJSON.version
+    rescriptJSWrapperPath = rescriptPackageJSON.bin.rescript
+  } catch (error) {
+    return null
+  }
+
+  let binaryPath: string | null = null
+  if (binary == "rescript") {
+    // Can't use the native bsb/rescript since we might need the watcher -w
+    // flag, which is only in the JS wrapper
+    binaryPath = path.join(rescriptDir, rescriptJSWrapperPath)
+  } else if (semver.gte(rescriptVersion, "12.0.0-alpha.13")) {
```

Comment:
AFAIK, npm's default semver comparison doesn't work intuitively with pre-releases. This might need the `includePrelease` flag.

https://www.npmjs.com/package/semver#prerelease-tags

Reply:
You're right that we'll need that flag for range matching, so I added it in another place where we're checking major version numbers:

https://github.com/rescript-lang/rescript-vscode/blob/22bff6d02b2c3ac35f8a6f86bd169ed16e8d3f87/server/src/incrementalCompilation.ts#L580

For `semver.gte(rescriptVersion, "12.0.0-alpha.13")` it won't be needed as we're not doing a range comparison.